File: python/tests/unit/dynamics/test_inverse_kinematics.py
import math
import platform

import dartpy as dart
import numpy as np
import pytest


class SsikLikeSolution:
    def __init__(self, q, validity=None):
        self.q = np.asarray(q)
        self.fk_residual = 0.0
        self.refinement_used = False
        if validity is not None:
            self.validity = validity


def _create_free_joint_ik():
    skel = dart.dynamics.Skeleton()
    [joint, body] = skel.createFreeJointAndBodyNodePair()
    ik = body.getOrCreateIK()
    dofs = list(range(skel.getNumDofs()))
    ik.setDofs(dofs)
    return skel, body, ik, dofs


def test_solve_for_free_joint():
    """
    Very simple test of InverseKinematics module, applied to a FreeJoint to
    ensure that the target is reachable
    """

    skel = dart.dynamics.Skeleton()
    [joint0, body0] = skel.createFreeJointAndBodyNodePair()

    ik = body0.getOrCreateIK()
    assert ik.isActive()

    tf = dart.math.Isometry3()
    tf.set_translation([0, 0, 0.8])
    tf.set_rotation(dart.math.AngleAxis(math.pi / 8.0, [0, 1, 0]).to_rotation_matrix())
    ik.getTarget().setTransform(tf)

    error_method = ik.getErrorMethod()
    assert error_method.getMethodName() == "TaskSpaceRegion"
    [lb, ub] = error_method.getBounds()
    assert len(lb) == 6
    assert len(ub) == 6
    error_method.setBounds(np.ones(6) * -1e-8, np.ones(6) * 1e-8)
    [lb, ub] = error_method.getBounds()
    assert lb == pytest.approx(-1e-8)
    assert ub == pytest.approx(1e-8)

    solver = ik.getSolver()
    solver.setNumMaxIterations(100)

    prob = ik.getProblem()

    tf_actual = ik.getTarget().getTransform().matrix()
    tf_expected = body0.getTransform().matrix()
    assert not np.isclose(tf_actual, tf_expected).all()

    success = solver.solve()
    assert success

    tf_actual = ik.getTarget().getTransform().matrix()
    tf_expected = body0.getTransform().matrix()
    assert np.isclose(tf_actual, tf_expected).all()


def test_python_analytical_accepts_ssik_like_solutions():
    skel, body, ik, dofs = _create_free_joint_ik()
    expected = np.linspace(0.1, 0.6, len(dofs))
    seen_targets = []

    def solve(target):
        seen_targets.append(np.asarray(target))
        return [SsikLikeSolution(expected)]

    analytical = ik.setPythonAnalytical(solve, dofs, "ssik.prebuilt.fake_ik")

    assert ik.getDofs() == dofs
    assert ik.getGradientMethod().getMethodName() == "ssik.prebuilt.fake_ik"
    assert ik.getAnalytical() is not None
    assert analytical.getDofs() == dofs

    target = dart.math.Isometry3()
    target.set_translation([0.1, 0.2, 0.3])

    solutions = analytical.getSolutions(target)

    assert len(solutions) == 1
    assert np.allclose(solutions[0].mConfig, expected)
    assert solutions[0].mValidity == dart.dynamics.InverseKinematicsAnalytical.VALID
    assert len(seen_targets) == 1
    assert seen_targets[0].shape == (4, 4)
    assert np.allclose(seen_targets[0], target.matrix())


def test_python_analytical_accepts_tuple_and_solution_objects():
    skel, body, ik, dofs = _create_free_joint_ik()
    valid = np.linspace(0.2, 0.7, len(dofs))
    out_of_reach = np.linspace(-0.2, -0.7, len(dofs))

    def solve(target):
        return [
            (out_of_reach, dart.dynamics.InverseKinematicsAnalytical.OUT_OF_REACH),
            dart.dynamics.InverseKinematicsAnalyticalSolution(valid),
        ]

    analytical = ik.setPythonAnalytical(solve, dofs)
    solutions = analytical.getSolutions(dart.math.Isometry3())

    assert len(solutions) == 2
    assert solutions[0].mValidity == dart.dynamics.InverseKinematicsAnalytical.VALID
    assert np.allclose(solutions[0].mConfig, valid)
    assert (
        solutions[1].mValidity == dart.dynamics.InverseKinematicsAnalytical.OUT_OF_REACH
    )
    assert np.allclose(solutions[1].mConfig, out_of_reach)


def test_python_analytical_can_leave_extra_dofs_to_dart_ik():
    skel, body, ik, dofs = _create_free_joint_ik()
    analytical_dofs = dofs[:3]

    def solve(target):
        return [np.zeros(len(analytical_dofs))]

    analytical = ik.setPythonAnalytical(solve, analytical_dofs, "partial_ssik")
    analytical.setExtraDofUtilization(
        dart.dynamics.InverseKinematicsAnalytical.PRE_AND_POST_ANALYTICAL
    )

    assert ik.getDofs() == dofs
    assert analytical.getDofs() == analytical_dofs
    assert ik.getAnalytical() is not None
    assert ik.getGradientMethod().getMethodName() == "partial_ssik"
    assert (
        analytical.getExtraDofUtilization()
        == dart.dynamics.InverseKinematicsAnalytical.PRE_AND_POST_ANALYTICAL
    )


def test_python_analytical_rejects_wrong_solution_dimension():
    skel, body, ik, dofs = _create_free_joint_ik()

    def solve(target):
        return [np.zeros(len(dofs) - 1)]

    analytical = ik.setPythonAnalytical(solve, dofs)

    with pytest.raises(ValueError, match="DOFs were registered"):
        analytical.getSolutions(dart.math.Isometry3())


class FailingSolver(dart.optimizer.Solver):
    def __init__(self, constant):
        super(FailingSolver, self).__init__()
        self.constant = constant

    def solve(self):
        problem = self.getProblem()
        if problem is None:
            print(
                "[FailingSolver::solve] Attempting to solve a nullptr problem! We will return false."
            )
            return False

        dim = problem.getDimension()
        wrong_solution = np.ones(dim) * self.constant
        problem.setOptimalSolution(wrong_solution)

        return False

    def getType(self):
        return "FailingSolver"

    def clone(self):
        return FailingSolver(self.constant)


def test_do_not_apply_solution_on_failure():
    skel = dart.dynamics.Skeleton()
    [joint, body] = skel.createFreeJointAndBodyNodePair()

    ik = body.getIK(True)
    solver = FailingSolver(10)
    ik.setSolver(solver)

    dofs = skel.getNumDofs()
    skel.resetPositions()

    assert not ik.solveAndApply(allowIncompleteResult=False)
    assert np.isclose(skel.getPositions(), np.zeros(dofs)).all()

    assert not ik.solveAndApply(allowIncompleteResult=True)
    assert not np.isclose(skel.getPositions(), np.zeros(dofs)).all()


if __name__ == "__main__":
    pytest.main()

File: tests/integration/optimization/test_inverse_kinematics.cpp
/*
 * Copyright (c) 2011, The DART development contributors
 * All rights reserved.
 *
 * The list of contributors can be found at:
 *   https://github.com/dartsim/dart/blob/main/LICENSE
 *
 * This file is provided under the following "BSD-style" License:
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted provided that the following
 *   conditions are met:
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *   POSSIBILITY OF SUCH DAMAGE.
 */

#include "helpers/gtest_utils.hpp"

#include "dart/config.hpp"
#include "dart/dynamics/All.hpp"
#include "dart/math/helpers.hpp"

#include <gtest/gtest.h>

#include <iostream>
#include <span>
#include <string_view>
#include <vector>

using namespace Eigen;
using namespace dart;
using namespace dart::dynamics;
using namespace dart::test;

namespace {

class DummyAnalyticalSolver final : public InverseKinematics::Analytical
{
public:
  explicit DummyAnalyticalSolver(
      InverseKinematics* ik, const Properties& properties = Properties())
    : Analytical(ik, "DummyAnalytical", properties),
      mDofs(ik->getDofs().begin(), ik->getDofs().end())
  {
    // Do nothing
  }

  std::unique_ptr<GradientMethod> clone(InverseKinematics* newIk) const override
  {
    return std::make_unique<DummyAnalyticalSolver>(
        newIk, getAnalyticalProperties());
  }

  std::span<const Solution> computeSolutions(const Eigen::Isometry3d&) override
  {
    mSolutions.clear();
    const int dofCount = static_cast<int>(mDofs.size());
    Eigen::VectorXd config = Eigen::VectorXd::Zero(dofCount);
    mSolutions.emplace_back(config, VALID);

    config.setConstant(0.1);
    mSolutions.emplace_back(config, OUT_OF_REACH);

    config.setConstant(-0.1);
    mSolutions.emplace_back(config, LIMIT_VIOLATED);

    checkSolutionJointLimits();

    return mSolutions;
  }

  std::span<const std::size_t> getDofs() const override
  {
    return mDofs;
  }

private:
  std::vector<std::size_t> mDofs;
};

} // namespace

//==============================================================================
TEST(InverseKinematics, SolveForFreeJoint)
{
  // Very simple test of InverseKinematics module, applied to a FreeJoint to
  // ensure that the target is reachable

  SkeletonPtr skel = Skeleton::create();
  skel->createJointAndBodyNodePair<FreeJoint>();

  std::shared_ptr<InverseKinematics> ik = skel->getBodyNode(0)->getIK(true);

  Eigen::Isometry3d tf(Eigen::Isometry3d::Identity());
  tf.translation() = Eigen::Vector3d(0.0, 0.0, 0.8);
  tf.rotate(Eigen::AngleAxisd(math::pi / 8, Eigen::Vector3d(0, 1, 0)));
  ik->getTarget()->setTransform(tf);

  ik->getErrorMethod().setBounds(
      Eigen::Vector6d::Constant(-1e-8), Eigen::Vector6d::Constant(1e-8));

  ik->getSolver()->setNumMaxIterations(100);

  EXPECT_FALSE(equals(
      ik->getTarget()->getTransform().matrix(),
      skel->getBodyNode(0)->getTransform().matrix(),
      1e-1));

  EXPECT_TRUE(ik->getSolver()->solve());

  EXPECT_TRUE(equals(
      ik->getTarget()->getTransform().matrix(),
      skel->getBodyNode(0)->getTransform().matrix(),
      1e-8));
}

//==============================================================================
class FailingSolver : public math::Solver
{
public:
  FailingSolver(double constant) : mConstant(constant)
  {
    // Do nothing
  }

  ~FailingSolver() override = default;

  bool solve() override
  {
    std::shared_ptr<math::Problem> problem = mProperties.mProblem;
    if (nullptr == problem) {
      DART_WARN("Attempting to solve a nullptr problem! We will return false.");
      return false;
    }

    const auto dim = problem->getDimension();
    const Eigen::VectorXd wrongSolution
        = Eigen::VectorXd::Constant(static_cast<int>(dim), mConstant);
    problem->setOptimalSolution(wrongSolution);

    return false;
  }

  std::string_view getType() const override
  {
    static constexpr std::string_view type = "FailingSolver";
    return type;
  }

  std::shared_ptr<Solver> clone() const override
  {
    return std::make_shared<FailingSolver>(mConstant);
  }

protected:
  double mConstant;
};

//==============================================================================
TEST(InverseKinematics, DoNotApplySolutionOnFailure)
{
  SkeletonPtr skel = Skeleton::create();
  skel->createJointAndBodyNodePair<FreeJoint>();

  std::shared_ptr<InverseKinematics> ik = skel->getBodyNode(0)->getIK(true);
  ik->setSolver(std::make_shared<FailingSolver>(10));
  ik->getTarget()->setTransform(Eigen::Isometry3d::Identity());

  const auto dofs = static_cast<int>(skel->getNumDofs());
  skel->resetPositions();

  EXPECT_FALSE(ik->solveAndApply(false));
  EXPECT_TRUE(equals(skel->getPositions(), Eigen::VectorXd::Zero(dofs).eval()));

  EXPECT_FALSE(ik->solveAndApply(true));
  EXPECT_FALSE(
      equals(skel->getPositions(), Eigen::VectorXd::Zero(dofs).eval()));
}

//==============================================================================
TEST(InverseKinematics, AnalyticalSortsSolutions)
{
  SkeletonPtr skel = Skeleton::create();
  skel->createJointAndBodyNodePair<FreeJoint>();

  std::shared_ptr<InverseKinematics> ik = skel->getBodyNode(0)->getIK(true);
  ik->setGradientMethod<DummyAnalyticalSolver>();

  auto* analytical = ik->getAnalytical();
  ASSERT_NE(analytical, nullptr);

  const auto solutions
      = analytical->getSolutions(Eigen::Isometry3d::Identity());
  EXPECT_EQ(solutions.size(), 3u);
}
